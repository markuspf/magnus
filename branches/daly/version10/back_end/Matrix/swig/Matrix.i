///created by xiaowei xu
////xuxw76@yahoo.com
///refer to word.i to see how swig works
/*
  for template class, we mapped ClassName<Type>  to ClassNameType
  such as: MatrixInteger,MatrixRational and so on.
  at the same time,the constructor name will changed to MatrixInteger,MatrixRational.
*/
%module Magnus
%include DArray.i
%{
#include "Matrix.h"
%}

%define LISTMATRIX(MATRIX,DARRAY,TYPE)
%{
typedef DArray<TYPE> DARRAY;
typedef Matrix<TYPE> MATRIX;
%}

class MATRIX:public DARRAY{

public:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Constructors:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////
  MATRIX(int n = 0 );
  // Makes Matrix with height=width=n


  %name(MATRIX ## FromInt) MATRIX(int h, int w);
  ///by xuxw:name equals to "MATRIXFromInt", such as MatrixIntegerFromInt,MatrixRationalFromInt.
  // Makes Matrix with height == h and width == w
  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Operators:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

//  Matrix& operator += ( const Matrix& M );

//  Matrix& operator -= ( const Matrix& M ) ;

//  Matrix& operator *= ( const Matrix& M ) ;
//  Matrix operator + ( const Matrix& M) const;

//  Matrix operator - ( const Matrix& M ) const;

//  Matrix operator * ( const Matrix& M ) const ;

//  Matrix operator - ( ) const;

  %addmethods {

	MATRIX __mul__(const MATRIX& w) {
		return (*self)*w;
	}

	MATRIX __add__(const MATRIX& w) {
		return (*self)+w;
	}

	MATRIX __sub__(const MATRIX& w) {
		return (*self)-w;
	}

  }

  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Accessors:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // I/O:                                                                //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  // Inherited from DArray  

  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // IPC tools:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  // Inherited from DArray  
/*
  protected :
	Matrix( MatrixRep<R> newrep ) : 
            DerivedObjectOf<DArray<R>,MatrixRep<R> >(new MatrixRep<R>(newrep)) {}
	Matrix( MatrixRep<R>* newrep ) : 
            DerivedObjectOf<DArray<R>,MatrixRep<R> >(newrep) {}
*/
};

%enddef
LISTMATRIX(MatrixInteger,DArrayInteger,Integer)
LISTMATRIX(MatrixRational,DArrayRational,Rational)
LISTMATRIX(Matrixdouble,DArraydouble,double)
