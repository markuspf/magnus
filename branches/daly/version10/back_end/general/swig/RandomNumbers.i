
%module Magnus

%{
#include "RandomNumbers.h"
%}

class UniformRandom {

// An object in this class generates uniformly random numbers of type
// int and float.

public:

//  UniformRandom(int seed = TimeSeed());
  UniformRandom(int seed);
  void reseed(int seed);
  // In particular, calling this with the same seed as passed to the
  // constructor starts the random sequence over.

  float rand( );
  // Returns uniformly random floats in [0.0, 1.0].

  %name(randFromInt) int rand(int lower, int upper);
  // Returns uniformly random ints in [lower, upper]

/*
  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // IPC tools:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  friend ostream& operator < ( ostream& ostr, const UniformRandom& unRand )
  {
    ostr < unRand.ix1 < unRand.ix2 < unRand.ix3;
    for (int i = 0; i<98; i++) { ostr << unRand.r[i] << ' ';}// There is no > and < 
    return ostr;                                             // operators for float in IPC.h
  }
  
  friend istream& operator > ( istream& istr, UniformRandom& unRand)
  {
    istr > unRand.ix1 > unRand.ix1 > unRand.ix2 > unRand.ix3;
    for (int i = 0; i<98; i++) { istr >> unRand.r[i];}
     return istr;
  }

private:

  static const int   M1  = 259200;
  static const int   IA1 = 7141;
  static const int   IC1 = 54773;
  static const float RM1 = (1.0/M1);
  static const int   M2  = 134456;
  static const int   IA2 = 8121;
  static const int   IC2 = 28411;
  static const float RM2 = (1.0/M2);
  static const int   M3  = 243000;
  static const int   IA3 = 4561;
  static const int   IC3 = 51349;

  long ix1, ix2, ix3;
  float r[98];
*/
};



class NormalRandom {

// An object in this class generates normally distributed random numbers
// of type int and float.

public:

//  NormalRandom(int seed = TimeSeed());
  NormalRandom(int seed);
  void reseed(int seed);
  // In particular, calling this with the same seed as passed to the
  // constructor starts the random sequence over.

  float rand( );
  // Returns normally distributed random floats with mean 0 and variance 1.

  %name(randFromInt) int rand(int mean, int stddev);
  // Returns normally distributed random ints with mean mean and standard
  // deviation stddev. For example, rand(100, 20) returns ints, roughly 68%
  // of which fall within 20 of 100, and 95% within 40 of 100.
  // Beware that rand(100, 20) could return a negative int, though it is
  // unlikely. You may want something like
  //   do { i = G.rand(100, 20); } while ( i < 1 );
/*
  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // IPC tools:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  friend ostream& operator < ( ostream& ostr, const NormalRandom& noRand )
  {
    ostr < noRand.uniformDeviates < noRand.iset;
    ostr << noRand.gset << ' ';
    return ostr;
  }
  
  friend istream& operator > ( istream& istr, NormalRandom& noRand)
  {
    istr > noRand.uniformDeviates > noRand.iset;
    istr >> noRand.gset;
    return istr;
  }

private:

  UniformRandom uniformDeviates;
  int iset;
  float gset;
*/
};
