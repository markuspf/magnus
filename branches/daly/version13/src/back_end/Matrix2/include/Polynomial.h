// Copyright (C) 1995 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.
//
// Contents: Definitions of classes Monomial, Polynomial.
//
// Principal Author: Dmitry Bormotov
//
// Status: In development
//
// Usage:
//
// Revision History:
//


#ifndef _POLYNOMIAL_H_
#define _POLYNOMIAL_H_

#include "Integer.h"
#include "Rational.h"
#include "RingParser.h"
#include "IStreamPoll.h"
#include "BTree.h"


//-------------------------------- Monomial ---------------------------------//


template <class R> class Monomial
{

  friend class Polynomial<R>;

public:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Constructors:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  Monomial( const char * );

  Monomial( R coef = 0, int numOfVars = 0, int *powersOfVars = NULL );

  Monomial( const Monomial& );

  ~Monomial( ) 
  {     
    if( numberOfVariables > 0 )
      delete [] powersOfVariables; 
  }


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Operators:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  bool operator == ( const Monomial& ) const;

  Monomial& operator = ( const Monomial& );

  Monomial operator - () const;

  Monomial operator * ( const Monomial& ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Accessors:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  int compare( const Monomial& ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // I/O:                                                                //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  friend ostream& operator << ( ostream& ostr, const Monomial& M )
  {
    M.printOn(ostr);
    return ostr;
  }

  friend IStreamPoll operator >> ( istream& istr, Monomial& M )
  {
    Chars errMesg;
    M = M.readFrom(istr, errMesg);
    return IStreamPoll( istr, errMesg );
  }


private:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Data Members:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  R coefficient;
  int numberOfVariables;
  int *powersOfVariables;


  //////////////////////////////////////////////////////////////
  //                                                          //
  // Private functions:                                       //
  //                                                          //
  //////////////////////////////////////////////////////////////

  void printOn( ostream& ) const;

  Monomial readFrom( istream& istr, Chars& errMesg )
  {
    MonomialParser<R> P(istr);
    return P.parseMonomial(errMesg);
  }


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  //  Debugging stuff:                                                   //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

#ifdef DEBUG

  friend int main( );

#endif

};


//------------------------------- Polynomial --------------------------------//


template <class R> 
class Polynomial : public BTree< Monomial<R> >
{

public:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Constructors:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  Polynomial( ) : BTree< Monomial<R> >( ) { }  

  Polynomial( const char*);

  Polynomial( const Monomial<R>& );

  Polynomial( const Polynomial& );


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Operators:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  bool operator == ( const Polynomial& ) const;

  Polynomial& operator = ( const Polynomial& );

  Polynomial& operator += ( const Polynomial& );

  Polynomial& operator -= ( const Polynomial& );

  Polynomial& operator *= ( const Polynomial& );

  Polynomial operator + ( const Polynomial& ) const; 

  Polynomial operator - ( const Polynomial& ) const; 

  Polynomial operator * ( const Polynomial& ) const; 

  Polynomial operator - ( ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Accessors:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  bool isIdentity( ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // I/O:                                                                //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  friend ostream& operator << ( ostream& ostr, const Polynomial& P )
  {
    P.printOn(ostr);
    return ostr;
  }

  friend IStreamPoll operator >> ( istream& istr, Polynomial& P )
  {
    Chars errMesg;
    P = P.readFrom(istr, errMesg);
    return IStreamPoll( istr, errMesg );
  }

private:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Data Members:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Private functions:                                                  //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  void printOn( ostream& ) const;

  Polynomial readFrom( istream& istr, Chars& errMesg )
  {
    PolynomialParser<R> P(istr);
    return P.parsePolynomial(errMesg);
  }
  

  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  //  Debugging stuff:                                                   //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

#ifdef DEBUG

  friend int main( );

#endif

};

#endif



