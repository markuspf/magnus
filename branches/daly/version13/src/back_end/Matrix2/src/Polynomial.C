// Copyright (C) 1995 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.
//
// Contents: Implementations of classes Monomial, Polynomial. 
//
// Principal Author: Dmitry Bormotov
//
// Status: In development
//
// Usage:
//
// Discussion:
//
// Revision History:
//
// Next Implementation Steps:
//


#include "memory.h"
#include "Polynomial.h"


//-------------------------------- Monomial ---------------------------------//


template <class R>
Monomial<R>::Monomial( const char *s )
{
  numberOfVariables = 0;
  istrstream istr(s);
  istr >> *this;
}


template <class R>
Monomial<R>::Monomial( R coef = 0, int numOfVars = 0,
		       int *powersOfVars = NULL ) :
    coefficient(coef), numberOfVariables(numOfVars)
{
#if SAFETY > 0
  if( numberOfVariables < 0 )
    error("Monomial<R>::Monomial( R, int, int * ) : "
	  "The number of variables can't be negative.");
//@db What about ( numberOfVariables > 0 && powersOfVars == NULL ) ?
#endif
  
  powersOfVariables = new int[numberOfVariables];
  memcpy( powersOfVariables, powersOfVars, numberOfVariables * sizeof(int) );
}


template <class R>
Monomial<R>::Monomial( const Monomial& M ) : coefficient(M.coefficient),
    numberOfVariables(M.numberOfVariables)
{
  powersOfVariables = new int[numberOfVariables];
  memcpy( powersOfVariables, M.powersOfVariables,
	  numberOfVariables * sizeof(int) );
}


template <class R>
Monomial<R>& Monomial<R>::operator = ( const Monomial& M )
{
  if( numberOfVariables != M.numberOfVariables ) {
    if( numberOfVariables > 0 )
      delete [] powersOfVariables;
    numberOfVariables = M.numberOfVariables;
    powersOfVariables = new int[numberOfVariables];
  }

  coefficient = M.coefficient;
  memcpy( powersOfVariables, M.powersOfVariables,
	  numberOfVariables * sizeof(int) );

  return *this;
}


template <class R>
bool Monomial<R>::operator == ( const Monomial& M ) const
{
  if( numberOfVariables != M.numberOfVariables ||
      coefficient != M.coefficient )
    return false;
  for( int i = 0; i < numberOfVariables; i++ )
    if( powersOfVariables[i] != M.powersOfVariables[i] )
      return false;
  return true;
}


template <class R>
Monomial<R> Monomial<R>::operator - () const
{
  Monomial<R> result(*this);
  result.coefficient = - result.coefficient;
  return result;
}


template <class R>
Monomial<R> Monomial<R>::operator * ( const Monomial& M ) const
{
  const Monomial *M1 = this;
  const Monomial *M2 = &M;
  if( M1->numberOfVariables < M2->numberOfVariables ) {
    const Monomial *R = M1;
    M1 = M2;
    M2 = R;
  }

  Monomial result(*M1);
  result.coefficient *= M2->coefficient;
  for( int i = 0; i < M2->numberOfVariables; ++i )
    result.powersOfVariables[i] += M2->powersOfVariables[i];

  return result;
}


template <class R>
int Monomial<R>::compare( const Monomial& M ) const
{
  if( numberOfVariables < M.numberOfVariables )
    return -1;
  else
    if( numberOfVariables > M.numberOfVariables )
      return 1;
  else 
    for( int i = 0; i < numberOfVariables; ++i ) {
      int num1 = powersOfVariables[i];
      int num2 = M.powersOfVariables[i];
      if( num1 < num2 )
	return -1;
      else
	if( num1 > num2 )
	  return 1;
    }

  return 0;
}


template <class R>
void Monomial<R>::printOn(ostream& ostr) const
{
  if( coefficient == -1)
    if( numberOfVariables == 0)
      ostr << " - 1";
    else 
      ostr << " -";
  else
    if( coefficient != 1 || numberOfVariables == 0)
      ostr << " " << coefficient;
  
  for( int i = 0; i < numberOfVariables; ++i ) {
    int power = powersOfVariables[i];
    if( power > 0 ) {
      ostr << " x" << i+1;
      if( power != 1 )
	ostr << '^' << power;
    }
  }
}


//------------------------------- Polynomial --------------------------------//


template <class R>
Polynomial<R>::Polynomial( const Polynomial& P ) : BTree< Monomial<R> > ( )
{
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    this->insert(I.value());
}


template <class R>
Polynomial<R>::Polynomial( const Monomial<R>& M ) : BTree< Monomial<R> > ( )
{
  this->insert(M);
}


template <class R>
Polynomial<R>::Polynomial( const char *s )
{
  istrstream istr(s);
  istr >> *this;
}


template <class R>
Polynomial<R>& Polynomial<R>::operator = ( const Polynomial& P )
{
  deleteAll();
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    this->insert(I.value());
  return *this;
}


template <class R>
bool Polynomial<R>::operator == ( const Polynomial& P ) const
{
  //@db ?????
}

template <class R>
Polynomial<R>& Polynomial<R>::operator += ( const Polynomial& P )
{
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    this->insert(I.value());
  return *this;
}


template <class R>
Polynomial<R>& Polynomial<R>::operator -= ( const Polynomial& P )
{
  (*this) += -P;
}


template <class R>
Polynomial<R>& Polynomial<R>::operator *= ( const Polynomial& P )
{
  Polynomial<R> result = (*this) * P;
  *this = result;
}


template <class R>
Polynomial<R> Polynomial<R>::operator + (const Polynomial& P) const
{
  Polynomial<R> result(*this);
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    result.insert(I.value());
  
  return result;
}


template <class R>
Polynomial<R> Polynomial<R>::operator - () const
{
  Polynomial<R> result;
  BTreeIterator< Monomial<R> > I(*this);
  for( ; !I.done(); I.next() )
    result.insert(-I.value());
}


template <class R>
Polynomial<R> Polynomial<R>::operator - (const Polynomial& P) const
{
  return (*this) + (-P);
}


template <class R>
Polynomial<R> Polynomial<R>::operator * (const Polynomial& P) const
{
  Polynomial<R> result;
  BTreeIterator< Monomial<R> > I(*this);
  BTreeIterator< Monomial<R> > J(P);
  for( ; !I.done(); I.next() )
    for( ; !J.done(); J.next() )
      result.insert( I.value() * J.value() );
  return result;
}


template <class R>
void Polynomial<R>::printOn(ostream& ostr) const
{
  bool bFirst = true;
  BTreeIterator< Monomial<R> > I(*this);
  for( ; !I.done(); I.next() ) {
    if( !bFirst && I.value().coefficient > 0 )
      cout << "+";
    cout << I.value() << " ";
    bFirst = false;
  }
  cout << endl;
}


template class Monomial<Integer>;
template class Monomial<Rational>;
template class Polynomial<Integer>;
template class Polynomial<Rational>;

