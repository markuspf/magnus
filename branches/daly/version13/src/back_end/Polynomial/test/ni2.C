// Copyright (C) 2000 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.
//
// Contents: Experimenting with the Newton's iteration
//
// Principal Authors: Dmitry Bormotov
//


//#include "double.h"
#include "WordParser.h"
#include "Matrix.h"
#include "Polynomial.h"
#include "Map.h"
#include "MatrixComputations.h"
#include "List.h"
#include "RandomNumbers.h"


Matrix<double> makeToeplitzMatrix( const VectorOf<double>& u, 
				     const VectorOf<double>& v )
  // contructs Toeplitz matrix from column u and row v, 
  // the first coordinate of v (0) is ignored.
{
  int n = u.length();
  Matrix<double> M(n);
  for( int i = 0; i < n; ++i )
    for( int j = i; j < n; ++j )
      M[j][j-i] = u[i];
  for( int i = 1; i < n; ++i )
    for( int j = i; j < n; ++j )
      M[j-i][j] = v[i];
  return M;
}

Matrix<double> multByConst( const Matrix<double>& M, double k )
{
  Matrix<double> R(M.width());

  for( int i = 0; i < M.width(); ++i )
    for( int j = 0; j < M.width(); ++j )
      R[i][j] = M[i][j] * k;
  
  return R;
}

template<class R> void printMatrix( const Matrix<R>& M )
{
  for ( int i = 0; i < M.height(); i++ ){

    cout << "| ";
    for ( int j = 0; j < M.width(); j++ )
      cout << double(M[i][j]) << ", ";

    if( i==M.height()-1 ) 
      cout << "|."<<endl;
    else 
      cout <<"|"<< endl;
  } 
} 

Matrix<double> transpose( const Matrix<double>& M )
{
  Matrix<double> R(M.width());

  for( int i = 0; i < M.width(); ++i )
    for( int j = 0; j < M.width(); ++j )
      R[i][j] = M[j][i];
  
  return R;
}


Matrix<double> makeRandomToeplitz( int size )
{
  const double max = 100;
  UniformRandom r(1),r2(2345);
  VectorOf<double> u(size), v(size);
  for( int i = 0; i < size; ++i ) {
    u[i] = max * r.rand();
    v[i] = max * r.rand(); 
    if( r2.rand() < .5 )
      u[i] *= -1;
    if( r2.rand() < .5 )
      v[i] *= -1;
  }
  return makeToeplitzMatrix(u,v);
}


Matrix<double> makeRandomMatrix( int size )
{
  const double max = 10;
  UniformRandom r(1),r2(2345);
  Matrix<double> R(size);

  for( int i = 0; i < size; ++i )
    for( int j = 0; j < size; ++j ) {
      R[i][j] =  max * r.rand();
      if( r2.rand() < .5 )
	R[i][j] *= -1;
    }

  return R;
}


int main( )
{
  Polynomial<Integer> p1;
  Monomial<Integer> m;
  Map Ma;

  cout.precision(10);
  const int numOfIter = 50;
  double tau = .1;

  const int size = 5;
  VectorOf<double> u(size), v(size);
  /*
  u[0] = 0.196;
  u[1] = 0.2;
  u[2] = 0.3;
  u[3] = 0.4;
  u[4] = 0.5;
  v[1] = 0.2;
  v[2] = 0.3;
  v[3] = 0.4;
  v[4] = 0.5;
  
  Matrix<double> M = makeToeplitzMatrix(u,v);
  */

  //Matrix<double> M = makeRandomToeplitz(size);
  Matrix<double> M = makeRandomMatrix(size);

  cout << "M: " << endl << M << endl;
  for( int i = 0; i < 1; ++i ) {
    M = M * transpose(M);
    cout << "M: " << endl << M << endl;
  }

  /*
  const int size = 2;
  Matrix<double> M(size);
  VectorOf<double> u(size), v(size);
  */
  /*
  M[0][0] = double(5,10);
  M[0][1] = double(4,10);
  M[1][0] = double(4,10);
  M[1][1] = double(5,10);
  */
  /*
  M[0][0] = .5;
  M[0][1] = .4;
  M[1][0] = .4;
  M[1][1] = .5;
  */

  //cout << "M: " << endl << M << endl;

  
  Matrix<Rational> MR(size);
  for( int i = 0; i < size; ++i )
    for( int j = 0; j < size; ++j )
      MR[i][j] = M[i][j];
  
  //cout << "MR: " << endl;
  //printMatrix(MR);
  //cout << endl;
  /*
  MatrixComputations<Rational> MC(MR);
  MC.invertMatrix();
  //cout << endl << MC.isInvertibleMatrix() << endl;
  if( MC.isInvertibleMatrix() != yes ) {
    cout << "The matrix is not invertible" << endl; 
    return 0;
  }

  Matrix<Rational> MI = MC.getInverseMatrix() ;
  cout << "Inverse: " << endl;
  printMatrix(MI);
  cout << endl;
  
  cout << "Checking M * M^-1 :" << endl << MR * MI << endl; 
  */
  for( int i = 0; i < size; ++i ) {
    u[i] = 0;
    v[i] = 0;
  }
  
  u[0] = 1;

  Matrix<double> I = makeToeplitzMatrix(u,v);
  
  //cout << "I: " << endl << I << endl;
  
  Matrix<double> I2 = I;
  for( int i = 0; i < size; ++i )
    I2[i][i] = 2;

  cout << "Newton's Iteration: " << endl;
  char c = 'y';
  int s = 0;
  //Matrix<double> Y = I;
  Matrix<double> Y(size);
  //Y[0][0] = double(5,10);
  //Y[0][1] = double(-4,10);
  //Y[1][0] = double(-4,10);
  //Y[1][1] = double(5,10);

  Y[0][0] = 1;
  Y[0][1] = 0;
  Y[1][0] = 0;
  Y[1][1] = 1;


  Matrix<double> X0(I), Mk(size);
  
  const double eps = 0.001, min = 1e-10, max = 1e10;
  ListOf<double> steps,ts;
  double t = 0, old_t = 0;
  
  while( true ) {

    double step = 1 - t;

    while( true ) {

      t += step;
      
      cout << "Trying step: " << step << "  t: " << t << endl << endl;
      
      Mk = I + multByConst(M - I, t);
      Y = X0;
      
      //cout << "Mk: " << Mk << endl;
      //cout << "X0: " << X0 << endl;
      
      while( true ) {

	for( int s = 0; s < 10; ++s ) {
	  Matrix<double> R1 = Mk * Y;
	  //cout << "R1: " << endl << R1 << endl;
	  Matrix<double> R2 = I2 - R1;
	  //cout << "R2: " << endl << R2 << endl;
	  Matrix<double> R = Y * R2;
	  //cout << "R: " << endl << R << endl;
	  Y = R;
	}
	Matrix<double> P =  Mk * Y;
	//cout << "The product M * M^-1: " << endl << P << endl;
	cout << "10 steps ... " << endl;
	 if( P[0][0] < min || P[0][0] > max || fabs(P[0][0] - 1) < min )
	  break;
      }
      
      //cout << "Approximating M^-1:" << endl << Y << endl;
      Matrix<double> P =  Mk * Y;
      //cout << "The product M * M^-1: " << endl << P << endl;
      if( P[0][0] < min || P[0][0] > max ) {
	cout << "Divergence!" << endl << endl;
	t -= step;
	step /= 2;
	continue;
      }
      else {
	cout << "Convergence!" << endl << endl;
	
	//cout << "fabs(t - 1): " << fabs(t - 1) << endl;
	//cout << "eps: " << eps << endl;

	if( fabs(t - 1) < eps || step < eps ) {
	  step = t - old_t;
	  cout << "Final step: " << step << "  t: " << t << endl << endl;
	  steps.append(step);
	  ts.append(t);
	  old_t = t;
	  break;
	}
	step /= 2;
	continue;
      }
    }
    
    if( fabs(t - 1) < eps )
      break;
    X0 = Y;  
  }

  cout << "Steps: " << steps << endl << "Tau_k: " << ts << endl;
}





