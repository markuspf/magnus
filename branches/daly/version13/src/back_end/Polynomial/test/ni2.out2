			General case:


1. 

Steps: -> 0.0068359375 -> 0.01406335831 -> 0.02772843791 ->
0.05620900597 -> 0.112332499 -> 0.224853611 -> 0.4495421377 ->
0.1084350126

Tau_k: -> 0.0068359375 -> 0.02089929581 -> 0.04862773372 ->
0.1048367397 -> 0.2171692387 -> 0.4420228497 -> 0.8915649874 -> 1


2.

Steps: -> 0.0009765625 -> 0.002439022064 -> 0.004866134841 ->
0.009926867555 -> 0.01965500387 -> 0.03958007445 -> 0.07838125109 ->
0.1574584384 -> 0.3145215885 -> 0.3721950567

Tau_k: -> 0.0009765625 -> 0.003415584564 -> 0.008281719405 ->
0.01820858696 -> 0.03786359083 -> 0.07744366528 -> 0.1558249164 ->
0.3132833548 -> 0.6278049433 -> 1


3.

Steps: -> 3.051757812e-05 -> 6.10332936e-05 -> 0.0001220591369 ->
0.0002440884741 -> 0.0004880577644 -> 0.0009756389099 ->
0.001964601746 -> 0.003921469312 -> 0.007812062797 -> 0.01562127212 ->
0.03121977887 -> 0.06248736857 -> 0.1249769165 -> 0.2497808799 ->
0.4993171178

Tau_k: -> 3.051757812e-05 -> 9.155087173e-05 -> 0.0002136100086 ->
0.0004576984827 -> 0.0009457562471 -> 0.001921395157 -> 0.003885996903
-> 0.007807466215 -> 0.01561952901 -> 0.03124080113 -> 0.06246058 ->
0.1249479486 -> 0.2499248651 -> 0.499705745 -> 0.9990228628

