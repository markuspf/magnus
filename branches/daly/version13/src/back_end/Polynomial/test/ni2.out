			Toeplitz case


1. cond = 13.8

Steps: -> 0.05126953125 -> 0.1023776531 -> 0.2050793871 ->
0.4101895466 -> 0.231083882

Tau_k: -> 0.05126953125 -> 0.1536471844 -> 0.3587265715 -> 0.768916118
-> 1


2. cond = 113.1

Steps: -> 0.0205078125 -> 0.04113101959 -> 0.08293133369 ->
0.1658230684 -> 0.3316712228 -> 0.357935543

Tau_k: -> 0.0205078125 -> 0.06163883209 -> 0.1445701658 ->
0.3103932342 -> 0.642064457 -> 1


3. cond = 10652.57

Steps: -> 0.000244140625 -> 0.0004881620407 -> 0.0009758473607 ->
0.002071650568 -> 0.004134702976 -> 0.008235084687 -> 0.0163334541 ->
0.03259700687 -> 0.06482355127 -> 0.1300046378 -> 0.2602788532 ->
0.4798129085

Tau_k: -> 0.000244140625 -> 0.0007323026657 -> 0.001708150026 ->
0.003779800594 -> 0.00791450357 -> 0.01614958826 -> 0.03248304236 ->
0.06508004923 -> 0.1299036005 -> 0.2599082383 -> 0.5201870915 -> 1


4. cond = .942 * 10^8

Steps: -> 1.192092896e-07 -> 2.384185507e-07 -> 4.768369877e-07 ->
9.536735206e-07 -> 1.907345222e-06 -> 3.814683168e-06 ->
7.629337233e-06 -> 1.525855805e-05 -> 3.051665045e-05 ->
6.103143831e-05 -> 0.0001220554265 -> 0.0002440810543 ->
0.0004880429284 -> 0.0009756092524 -> 0.001951216644 -> 0.003898603713
-> 0.007781829844 -> 0.01550167765 -> 0.03075169479 -> 0.06138372778
-> 0.1224408893 -> 0.245307385 -> 0.4901414093 -> 0.01888983119

Tau_k: -> 1.192092896e-07 -> 3.576278402e-07 -> 8.344648279e-07 ->
1.788138349e-06 -> 3.695483571e-06 -> 7.510166739e-06 ->
1.513950397e-05 -> 3.039806202e-05 -> 6.091471247e-05 ->
0.0001219461508 -> 0.0002440015773 -> 0.0004880826316 -> 0.00097612556
-> 0.001951734812 -> 0.003902951456 -> 0.00780155517 -> 0.01558338501
-> 0.03108506267 -> 0.06183675745 -> 0.1232204852 -> 0.2456613745 ->
0.4909687596 -> 0.9811101688 -> 1


